src/UCTSearch.cpp

```diff
@@ -218,6 +217,69 @@ bool UCTSearch::is_running() const {
     return m_run && m_nodes < MAX_TREE_SIZE;
 }
 
+int UCTSearch::est_playouts_left() const {
+    auto elapsed_millis = now() - m_start_time;
+    auto playouts = m_playouts.load();
+    if (m_target_time < 0) {
+        // No time control, use playouts or visits.
+        const auto playouts_left = std::min(m_maxplayouts - playouts,
+                                            m_maxvisits - m_root->get_visits());
+        return playouts_left;
+    } else if (elapsed_millis < 1000 || playouts < 100) {
+        // Until we reach 1 second or 100 playouts playout_rate
+        // is not reliable, so just return max.
+        return MAXINT_DIV2;
+    } else {
+        const auto playout_rate = 1.0f * playouts / elapsed_millis;
+        const auto time_left = m_target_time - elapsed_millis;
+        return static_cast<int>(std::ceil(playout_rate * time_left));
+    }
+}
+
+size_t UCTSearch::prune_noncontenders() {
+    auto Nfirst = 0;
+    for (const auto& node : m_root->get_children()) {
+        Nfirst = std::max(Nfirst, node->get_visits());
+    }
+    const auto min_required_visits =
+        Nfirst - est_playouts_left();
+    auto pruned_nodes = size_t{0};
+    for (const auto& node : m_root->get_children()) {
+        const auto has_enough_visits =
+            node->get_visits() >= min_required_visits;
+        node->set_active(has_enough_visits);
+        if (!has_enough_visits) {
+            ++pruned_nodes;
+        }
+    }
+
+    assert(pruned_nodes < m_root->get_children().size());
+    return pruned_nodes;
+}
+
+bool UCTSearch::have_alternate_moves() {
+    if (!cfg_timemanage) {
+        // When timemanage is off always return true unless
+        // there is only one legal move.
+        return m_root->get_children().size() > 1;
```

Comment:
Is this going to affect training? E.g., forgoing the full 800 playouts will presumably make root_uct_winrate less accurate for positions with a single legal move. Will that be a problem?

Reply:
The dump_training outputs are normalized, so it used to be 800/800 = 100% on the only legal move. Now it will have 1/1 = 100% on the only legal move, so there is no difference.

Reply:
Check Training::record for the normalization code.

Reply:
I do see the normalization of move probabilities, but what about root_uct_winrate?

Reply:
Hey while double checking this I noticed #220!

Reply:
Oh, I didn't think of that, thanks! I will change it back. Wow that would have been a big mistake!